File: backend/app/schemas/team.py
"""
=========================================================================
Team Schemas - Pydantic Models for API Validation
SDLC Orchestrator - Sprint 70 (Teams Foundation)

Version: 1.0.0
Date: January 20, 2026
Status: ACTIVE - Sprint 70 Implementation
Authority: Backend Lead + CTO Approved
Reference: ADR-028-Teams-Feature-Architecture
Reference: Teams-Data-Model-Specification.md

Purpose:
- API request/response validation for Organizations, Teams, TeamMembers
- SASE-compliant role validation (ai_agent cannot be owner/admin)
- Type-safe data transfer between API and business logic

SDLC 5.1.2 Alignment:
- RoleEnum with ai_agent role for SE4A
- MemberTypeEnum with human/ai_agent types
- Settings schemas for SASE artifacts

Zero Mock Policy: Production-ready Pydantic v2 models
=========================================================================
"""

from datetime import datetime
from enum import Enum
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator


# =========================================================================
# Enums
# =========================================================================

class OrganizationPlan(str, Enum):
    """Organization subscription plan."""
    FREE = "free"
    STARTER = "starter"
    PRO = "pro"
    ENTERPRISE = "enterprise"


class TeamRole(str, Enum):
    """Team member role per SASE framework."""
    OWNER = "owner"       # SE4H with full VCR authority
    ADMIN = "admin"       # SE4H for member/settings management
    MEMBER = "member"     # SE4H/SE4A for implementation work
    AI_AGENT = "ai_agent" # SE4A (Agent Executor) for autonomous tasks


class MemberType(str, Enum):
    """Member type per SASE framework (CTO R2)."""
    HUMAN = "human"       # SE4H (Agent Coach or Member)
    AI_AGENT = "ai_agent" # SE4A (Agent Executor)


class AgenticMaturity(str, Enum):
    """SASE agentic maturity level."""
    L0 = "L0"  # Manual only
    L1 = "L1"  # AI-assisted
    L2 = "L2"  # AI-delegated
    L3 = "L3"  # AI-autonomous


# =========================================================================
# Organization Schemas
# =========================================================================

class OrganizationSettings(BaseModel):
    """Organization-wide settings including SASE config."""
    default_policy_pack: Optional[str] = Field(
        None,
        description="Default policy pack for new projects"
    )
    require_mfa: bool = Field(
        False,
        description="Require MFA for all organization members"
    )
    allowed_domains: list[str] = Field(
        default_factory=list,
        description="Allowed email domains for signup"
    )
    max_teams: Optional[int] = Field(
        None,
        ge=1,
        description="Maximum number of teams (NULL = unlimited)"
    )
    max_projects_per_team: Optional[int] = Field(
        None,
        ge=1,
        description="Maximum projects per team (NULL = unlimited)"
    )
    sase_config: Optional[dict] = Field(
        default_factory=dict,
        description="SASE configuration (agentic_maturity, mentor_scripts, etc.)"
    )

    model_config = ConfigDict(from_attributes=True)


class OrganizationCreate(BaseModel):
    """Schema for creating a new organization."""
    name: str = Field(
        ...,
        min_length=2,
        max_length=255,
        description="Display name of the organization",
        examples=["Acme Corporation"]
    )
    slug: str = Field(
        ...,
        min_length=2,
        max_length=100,
        pattern=r"^[a-z0-9]+(?:-[a-z0-9]+)*$",
        description="URL-safe unique identifier",
        examples=["acme-corp"]
    )
    plan: OrganizationPlan = Field(
        OrganizationPlan.FREE,
        description="Subscription plan"
    )
    settings: Optional[OrganizationSettings] = Field(
        default_factory=OrganizationSettings,
        description="Organization-wide settings"
    )

    model_config = ConfigDict(from_attributes=True)


class OrganizationUpdate(BaseModel):
    """Schema for updating an organization."""
    name: Optional[str] = Field(
        None,
        min_length=2,
        max_length=255,
        description="Display name of the organization"
    )
    slug: Optional[str] = Field(
        None,
        min_length=2,
        max_length=100,
        pattern=r"^[a-z0-9]+(?:-[a-z0-9]+)*$",
        description="URL-safe unique identifier"
    )
    plan: Optional[OrganizationPlan] = Field(
        None,
        description="Subscription plan"
    )
    settings: Optional[OrganizationSettings] = Field(
        None,
        description="Organization-wide settings"
    )

    model_config = ConfigDict(from_attributes=True)


class OrganizationResponse(BaseModel):
    """Schema for organization response."""
    id: UUID = Field(..., description="Organization UUID")
    name: str = Field(..., description="Display name")
    slug: str = Field(..., description="URL-safe identifier")
    plan: OrganizationPlan = Field(..., description="Subscription plan")
    settings: dict = Field(default_factory=dict, description="Organization settings")
    teams_count: int = Field(0, description="Number of teams")
    users_count: int = Field(0, description="Number of users")
    created_at: datetime = Field(..., description="Creation timestamp")
    updated_at: datetime = Field(..., description="Last update timestamp")

    model_config = ConfigDict(from_attributes=True)


class OrganizationListResponse(BaseModel):
    """Schema for paginated organization list."""
    items: list[OrganizationResponse] = Field(default_factory=list)
    total: int = Field(0, ge=0)
    page: int = Field(1, ge=1)
    page_size: int = Field(20, ge=1, le=100)
    has_next: bool = Field(False)

    model_config = ConfigDict(from_attributes=True)


# =========================================================================
# Team Schemas
# =========================================================================

class TeamSettings(BaseModel):
    """Team-specific settings including SASE config."""
    default_gate_approvers: list[UUID] = Field(
        default_factory=list,
        description="Default approvers for gate reviews"
    )
    notification_channel: Optional[str] = Field(
        None,
        pattern=r"^(slack|email|webhook)$",
        description="Notification channel type"
    )
    webhook_url: Optional[str] = Field(
        None,
        max_length=500,
        description="Webhook URL for notifications"
    )
    auto_assign_projects: bool = Field(
        False,
        description="Auto-assign new projects to team"
    )
    mentor_scripts: list[str] = Field(
        default_factory=list,
        description="MentorScript references"
    )
    briefing_templates: list[str] = Field(
        default_factory=list,
        description="BriefingScript template references"
    )
    agentic_maturity: AgenticMaturity = Field(
        AgenticMaturity.L0,
        description="SASE agentic maturity level"
    )
    crp_threshold: float = Field(
        0.7,
        ge=0.0,
        le=1.0,
        description="CRP threshold (confidence below triggers consultation)"
    )
    auto_approve_mrp: bool = Field(
        False,
        description="Auto-approve MRP (no human VCR required)"
    )

    model_config = ConfigDict(from_attributes=True)


class TeamCreate(BaseModel):
    """Schema for creating a new team."""
    organization_id: UUID = Field(..., description="Parent organization UUID")
    name: str = Field(
        ...,
        min_length=2,
        max_length=255,
        description="Display name of the team",
        examples=["Backend Team"]
    )
    slug: str = Field(
        ...,
        min_length=2,
        max_length=100,
        pattern=r"^[a-z0-9]+(?:-[a-z0-9]+)*$",
        description="URL-safe identifier (unique per organization)",
        examples=["backend-team"]
    )
    description: Optional[str] = Field(
        None,
        max_length=1000,
        description="Team description"
    )
    settings: Optional[TeamSettings] = Field(
        default_factory=TeamSettings,
        description="Team-specific settings"
    )

    model_config = ConfigDict(from_attributes=True)


class TeamUpdate(BaseModel):
    """Schema for updating a team."""
    name: Optional[str] = Field(
        None,
        min_length=2,
        max_length=255,
        description="Display name of the team"
    )
    slug: Optional[str] = Field(
        None,
        min_length=2,
        max_length=100,
        pattern=r"^[a-z0-9]+(?:-[a-z0-9]+)*$",
        description="URL-safe identifier"
    )
    description: Optional[str] = Field(
        None,
        max_length=1000,
        description="Team description"
    )
    settings: Optional[TeamSettings] = Field(
        None,
        description="Team-specific settings"
    )

    model_config = ConfigDict(from_attributes=True)


class TeamResponse(BaseModel):
    """Schema for team response."""
    id: UUID = Field(..., description="Team UUID")
    organization_id: UUID = Field(..., description="Parent organization UUID")
    name: str = Field(..., description="Display name")
    slug: str = Field(..., description="URL-safe identifier")
    description: Optional[str] = Field(None, description="Team description")
    settings: dict = Field(default_factory=dict, description="Team settings")
    members_count: int = Field(0, description="Number of members")
    human_members_count: int = Field(0, description="Number of human members")
    ai_agents_count: int = Field(0, description="Number of AI agents")
    projects_count: int = Field(0, description="Number of projects")
    created_at: datetime = Field(..., description="Creation timestamp")
    updated_at: datetime = Field(..., description="Last update timestamp")

    model_config = ConfigDict(from_attributes=True)


class TeamListResponse(BaseModel):
    """Schema for paginated team list."""
    items: list[TeamResponse] = Field(default_factory=list)
    total: int = Field(0, ge=0)
    page: int = Field(1, ge=1)
    page_size: int = Field(20, ge=1, le=100)
    has_next: bool = Field(False)

    model_config = ConfigDict(from_attributes=True)


class TeamStatistics(BaseModel):
    """Schema for team statistics."""
    team_id: UUID = Field(..., description="Team UUID")
    team_name: str = Field(..., description="Team display name")
    total_members: int = Field(0, description="Total members")
    human_members: int = Field(0, description="Human members (SE4H)")
    ai_agents: int = Field(0, description="AI agents (SE4A)")
    owners_count: int = Field(0, description="Number of owners")
    admins_count: int = Field(0, description="Number of admins")
    total_projects: int = Field(0, description="Total projects")
    active_projects: int = Field(0, description="Active projects")
    agentic_maturity: AgenticMaturity = Field(
        AgenticMaturity.L0,
        description="Current agentic maturity level"
    )

    model_config = ConfigDict(from_attributes=True)


# =========================================================================
# TeamMember Schemas
# =========================================================================

class TeamMemberAdd(BaseModel):
    """Schema for adding a member to a team."""
    team_id: UUID = Field(..., description="Team UUID")
    user_id: UUID = Field(..., description="User UUID to add")
    role: TeamRole = Field(
        TeamRole.MEMBER,
        description="Member role in team"
    )
    member_type: MemberType = Field(
        MemberType.HUMAN,
        description="Member type (human or ai_agent)"
    )

    @model_validator(mode="after")
    def validate_ai_agent_role(self) -> "TeamMemberAdd":
        """Validate that AI agents cannot have owner/admin roles (CTO R1/R2)."""
        if self.member_type == MemberType.AI_AGENT and self.role in (TeamRole.OWNER, TeamRole.ADMIN):
            raise ValueError(
                "AI agents cannot be owners or admins per SASE framework. "
                "Only humans (SE4H) can have VCR authority."
            )
        return self

    model_config = ConfigDict(from_attributes=True)


class TeamMemberUpdate(BaseModel):
    """Schema for updating a team member."""
    role: Optional[TeamRole] = Field(
        None,
        description="New role in team"
    )

    model_config = ConfigDict(from_attributes=True)


class TeamMemberResponse(BaseModel):
    """Schema for team member response."""
    id: UUID = Field(..., description="Membership UUID")
    team_id: UUID = Field(..., description="Team UUID")
    user_id: UUID = Field(..., description="User UUID")
    role: TeamRole = Field(..., description="Member role")
    member_type: MemberType = Field(..., description="Member type")
    sase_role: str = Field(..., description="SASE framework role designation")
    joined_at: datetime = Field(..., description="When user joined the team")
    created_at: datetime = Field(..., description="Creation timestamp")
    updated_at: datetime = Field(..., description="Last update timestamp")

    # User details (populated from relationship)
    user_email: Optional[str] = Field(None, description="User email")
    user_name: Optional[str] = Field(None, description="User display name")
    user_avatar_url: Optional[str] = Field(None, description="User avatar URL")

    model_config = ConfigDict(from_attributes=True)


class TeamMemberListResponse(BaseModel):
    """Schema for paginated team member list."""
    items: list[TeamMemberResponse] = Field(default_factory=list)
    total: int = Field(0, ge=0)
    page: int = Field(1, ge=1)
    page_size: int = Field(20, ge=1, le=100)
    has_next: bool = Field(False)

    model_config = ConfigDict(from_attributes=True)


# =========================================================================
# User Assignment Schemas
# =========================================================================

class UserOrganizationAssign(BaseModel):
    """Schema for assigning a user to an organization."""
    user_id: UUID = Field(..., description="User UUID")
    organization_id: UUID = Field(..., description="Organization UUID")

    model_config = ConfigDict(from_attributes=True)


class ProjectTeamAssign(BaseModel):
    """Schema for assigning a project to a team."""
    project_id: UUID = Field(..., description="Project UUID")
    team_id: UUID = Field(..., description="Team UUID")

    model_config = ConfigDict(from_attributes=True)


# =========================================================================
# Bulk Operation Schemas
# =========================================================================

class BulkAddMembersRequest(BaseModel):
    """Schema for bulk adding members to a team."""
    team_id: UUID = Field(..., description="Team UUID")
    members: list[TeamMemberAdd] = Field(
        ...,
        min_length=1,
        max_length=100,
        description="List of members to add"
    )

    model_config = ConfigDict(from_attributes=True)


class BulkOperationResult(BaseModel):
    """Schema for bulk operation result."""
    success_count: int = Field(0, description="Number of successful operations")
    failure_count: int = Field(0, description="Number of failed operations")
    errors: list[dict] = Field(
        default_factory=list,
        description="List of errors with details"
    )

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/lib/types/team.ts
/**
 * Team & TeamMember Types - Next.js App Router
 * @module frontend/src/lib/types/team
 * @status Sprint 84 - Teams & Organizations UI
 * @description Type definitions for Teams and TeamMember management
 * @note Based on backend schemas from Sprint 71 (Teams Backend API)
 * @see backend/app/schemas/team.py
 */

// =========================================================================
// Team Roles & Member Types (SASE Compliance)
// =========================================================================

/**
 * Team member roles per SASE specification
 * - owner: SE4H Coach (VCR authority, full control)
 * - admin: SE4H Coach (member management, limited control)
 * - member: SE4H Member (implementation)
 * - ai_agent: SE4A Executor (autonomous tasks)
 */
export type TeamRole = "owner" | "admin" | "member" | "ai_agent";

/**
 * Member types for SASE constraints
 * - AI agents cannot be owner or admin (CTO R1/R2)
 */
export type MemberType = "human" | "ai_agent";

/**
 * SASE role mapping for display
 */
export type SASERole = "SE4H_Coach" | "SE4H_Member" | "SE4A_Executor";

/**
 * Agentic maturity levels (L0-L3)
 */
export type AgenticMaturity = "L0" | "L1" | "L2" | "L3";

// =========================================================================
// Team Settings
// =========================================================================

export interface TeamSASEConfig {
  agentic_maturity?: AgenticMaturity;
  mentor_scripts?: string[];
  briefing_templates?: string[];
  crp_threshold?: number; // 0.0 - 1.0, default: 0.7
  auto_approve_mrp?: boolean;
}

export interface TeamSettings {
  default_gate_approvers?: string[];
  notification_channel?: "slack" | "email" | "webhook" | null;
  webhook_url?: string | null;
  auto_assign_projects?: boolean;
  mentor_scripts?: string[];
  briefing_templates?: string[];
  agentic_maturity?: AgenticMaturity;
  crp_threshold?: number;
  auto_approve_mrp?: boolean;
}

// =========================================================================
// Team Model
// =========================================================================

export interface Team {
  id: string;
  organization_id: string;
  name: string;
  slug: string;
  description: string | null;
  settings: TeamSettings;
  members_count: number;
  human_members_count: number;
  ai_agents_count: number;
  projects_count: number;
  created_at: string;
  updated_at: string;
}

export interface TeamDetail extends Team {
  organization_name?: string;
  agentic_maturity: AgenticMaturity;
  crp_threshold: number;
  auto_approve_mrp: boolean;
}

// =========================================================================
// Team CRUD Requests
// =========================================================================

export interface TeamCreate {
  organization_id: string;
  name: string;
  slug: string;
  description?: string | null;
  settings?: TeamSettings;
}

export interface TeamUpdate {
  name?: string;
  description?: string | null;
  settings?: Partial<TeamSettings>;
}

// =========================================================================
// Team Statistics
// =========================================================================

export interface TeamStatistics {
  team_id: string;
  team_name: string;
  total_members: number;
  human_members: number;
  ai_agents: number;
  owners_count: number;
  admins_count: number;
  total_projects: number;
  active_projects: number;
  agentic_maturity: AgenticMaturity;
}

// =========================================================================
// TeamMember Model
// =========================================================================

export interface TeamMember {
  id: string;
  team_id: string;
  user_id: string;
  role: TeamRole;
  member_type: MemberType;
  sase_role: SASERole;
  joined_at: string | null;
  created_at: string;
  updated_at: string;
  user_email: string | null;
  user_name: string | null;
  user_avatar_url: string | null;
}

export interface TeamMemberWithPermissions extends TeamMember {
  is_owner: boolean;
  is_admin: boolean;
  is_admin_or_owner: boolean;
  is_ai_agent: boolean;
  is_human: boolean;
  is_coach: boolean;
  is_executor: boolean;
  can_manage_members: boolean;
  can_approve_vcr: boolean;
  can_modify_settings: boolean;
  can_approve_sprint_gate: boolean;
  can_create_sprint: boolean;
  can_manage_backlog: boolean;
}

// =========================================================================
// TeamMember CRUD Requests
// =========================================================================

export interface TeamMemberAdd {
  user_id: string;
  role?: TeamRole;
  member_type?: MemberType;
}

export interface TeamMemberRoleUpdate {
  role: TeamRole;
}

// =========================================================================
// Paginated Responses
// =========================================================================

export interface PaginatedResponse<T> {
  items: T[];
  total: number;
  page: number;
  page_size: number;
  has_next: boolean;
}

export type TeamListResponse = PaginatedResponse<Team>;
export type TeamMemberListResponse = PaginatedResponse<TeamMember>;

// =========================================================================
// Query Parameters
// =========================================================================

export interface TeamListParams {
  skip?: number;
  limit?: number;
  organization_id?: string;
}

export interface TeamMemberListParams {
  skip?: number;
  limit?: number;
}

// =========================================================================
// Role Metadata for UI
// =========================================================================

export const TEAM_ROLE_META: Record<TeamRole, { label: string; color: string; description: string }> = {
  owner: {
    label: "Owner",
    color: "bg-purple-100 text-purple-800",
    description: "Full control, VCR authority (SE4H Coach)",
  },
  admin: {
    label: "Admin",
    color: "bg-blue-100 text-blue-800",
    description: "Member management, gate approval (SE4H Coach)",
  },
  member: {
    label: "Member",
    color: "bg-green-100 text-green-800",
    description: "Implementation tasks (SE4H Member)",
  },
  ai_agent: {
    label: "AI Agent",
    color: "bg-orange-100 text-orange-800",
    description: "Autonomous execution (SE4A Executor)",
  },
};

export const MEMBER_TYPE_META: Record<MemberType, { label: string; color: string }> = {
  human: { label: "Human", color: "bg-green-100 text-green-800" },
  ai_agent: { label: "AI Agent", color: "bg-orange-100 text-orange-800" },
};

export const SASE_ROLE_META: Record<SASERole, { label: string; description: string }> = {
  SE4H_Coach: {
    label: "SE4H Coach",
    description: "Human coach with governance authority",
  },
  SE4H_Member: {
    label: "SE4H Member",
    description: "Human team member",
  },
  SE4A_Executor: {
    label: "SE4A Executor",
    description: "AI agent executor",
  },
};

export const AGENTIC_MATURITY_META: Record<AgenticMaturity, { label: string; description: string; color: string }> = {
  L0: {
    label: "L0 - Manual",
    description: "No AI agents, fully human team",
    color: "bg-gray-100 text-gray-800",
  },
  L1: {
    label: "L1 - Assisted",
    description: "AI assists with suggestions, human reviews all",
    color: "bg-blue-100 text-blue-800",
  },
  L2: {
    label: "L2 - Collaborative",
    description: "AI and humans collaborate, AI can execute low-risk tasks",
    color: "bg-yellow-100 text-yellow-800",
  },
  L3: {
    label: "L3 - Autonomous",
    description: "AI operates autonomously within defined boundaries",
    color: "bg-green-100 text-green-800",
  },
};

// =========================================================================
// Helper Functions
// =========================================================================

/**
 * Get SASE role from team role and member type
 */
export function getSASERole(role: TeamRole, memberType: MemberType): SASERole {
  if (memberType === "ai_agent") return "SE4A_Executor";
  if (role === "owner" || role === "admin") return "SE4H_Coach";
  return "SE4H_Member";
}

/**
 * Check if role change is valid per SASE constraints
 * AI agents cannot be owner or admin
 */
export function isValidRoleAssignment(role: TeamRole, memberType: MemberType): boolean {
  if (memberType === "ai_agent" && (role === "owner" || role === "admin")) {
    return false;
  }
  return true;
}

/**
 * Format team slug from name
 */
export function formatTeamSlug(name: string): string {
  return name
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, "-")
    .replace(/^-+|-+$/g, "")
    .substring(0, 100);
}

/**
 * Get role badge color class
 */
export function getRoleBadgeClass(role: TeamRole): string {
  return TEAM_ROLE_META[role]?.color ?? "bg-gray-100 text-gray-800";
}

/**
 * Get member type icon
 */
export function getMemberTypeIcon(memberType: MemberType): string {
  return memberType === "ai_agent" ? "Bot" : "User";
}
